lib/galaxy/celery/__init__.py

```diff
@@ -67,6 +69,21 @@ def trim_module_name(self, module):
         return module
 
 
+class GalaxyTask(Task):
+    """
+    Custom celery task used to limit number of tasks executions per user
+    per second.
+    """
+
+    def before_start(self, task_id, args, kwargs):
+        """
+        Set appropriate before start object from DI container.
+        """
+        app = get_galaxy_app()
+        assert app
+        app[GalaxyTaskBeforeStart](self, task_id, args, kwargs)
```

Comment:
I think it would be more intuitive if you called a method here instead of the object itself. Consider this:
```py
pretask = app[GalaxyTaskBeforeStart]
pretask.check_rate_limit(self, kwargs)   # Also, you don't need task_id and kwargs for that method
```
I think an explicit call to a method with a descriptive name makes it immediately clear what's going on. As a side benefit, it offers flexibility: we can call other methods if we need to. (For the base object, it would be a noop, just like the `__call__` implementation)

Reply:
This was a deliberate design decision. The GalaxyTaskBeforeStart was intended as the class that would have all of the before start logic. If we needed to add more functionality it would be done within the GalaxyTaskBeforeStart.__call__ method without having to change the actual GalaxyTask.before_start method. So we either place the calls to the lower level distinct operations in the GalaxyTask.before_start or in GalaxyTaskBeforeStart.__call__ method. Because we might add other logic in the future I decided to pass all available information including the task_id and args. I see this as design preference.